url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/drafts/Draft_10.java

```diff
@@ -0,0 +1,397 @@
+		byte[] buffer = new byte[ bytecount ];
+		int highest = 8 * bytecount - 8;
+		for( int i = 0 ; i < bytecount ; i++ ) {
+			buffer[ i ] = (byte) ( val >>> ( highest - 8 * i ) );
+		}
+		return buffer;
+	}
+
+	private Opcode toOpcode( byte opcode ) throws InvalidFrameException {
+		switch ( opcode ) {
+			case 0:
+				return Opcode.CONTINUOUS;
+			case 1:
+				return Opcode.TEXT;
+			case 2:
+				return Opcode.BINARY;
+				// 3-7 are not yet defined
+			case 8:
+				return Opcode.CLOSING;
+			case 9:
+				return Opcode.PING;
+			case 10:
+				return Opcode.PONG;
+				// 11-15 are not yet defined
+			default :
+				throw new InvalidFrameException( "unknow optcode " + (short) opcode );
+		}
+	}
+
+	@Override
+	public List<Framedata> translateFrame( ByteBuffer buffer ) throws LimitExedeedException , InvalidDataException {
+		List<Framedata> frames = new LinkedList<Framedata>();
+		Framedata cur;
+
+		if( incompleteframe != null ) {
+			// complete an incomplete frame
+			while ( true ) {
+				try {
+					buffer.mark();
+					int available_next_byte_count = buffer.remaining();// The number of bytes received
+					int expected_next_byte_count = incompleteframe.remaining();// The number of bytes to complete the incomplete frame
+
+					if( expected_next_byte_count > available_next_byte_count ) {
+						// did not receive enough bytes to complete the frame
+						incompleteframe.put( buffer.array(), buffer.position(), available_next_byte_count );
+						buffer.position( buffer.position() + available_next_byte_count );
+						return Collections.emptyList();
+					}
+					incompleteframe.put( buffer.array(), buffer.position(), expected_next_byte_count );
+					buffer.position( buffer.position() + expected_next_byte_count );
+
+					cur = translateSingleFrame( (ByteBuffer) incompleteframe.duplicate().position( 0 ) );
+					frames.add( cur );
+					incompleteframe = null;
+					break; // go on with the normal frame receival
+				} catch ( IncompleteException e ) {
+					// extending as much as suggested
+					int oldsize = incompleteframe.limit();
+					ByteBuffer extendedframe = ByteBuffer.allocate( checkAlloc( e.getPreferedSize() ) );
+					assert ( extendedframe.limit() > incompleteframe.limit() );
+					incompleteframe.rewind();
+					extendedframe.put( incompleteframe );
+					incompleteframe = extendedframe;
+					
+					return translateFrame( buffer );
+				}
+			}
+		}
+
+		while ( buffer.hasRemaining() ) {// Read as much as possible full frames
+			buffer.mark();
+			try {
+				cur = translateSingleFrame( buffer );
+				frames.add( cur );
+			} catch ( IncompleteException e ) {
+				// remember the incomplete data
+				buffer.reset();
+				int pref = e.getPreferedSize();
+				incompleteframe = ByteBuffer.allocate( checkAlloc( pref ) );
+				incompleteframe.put( buffer );
+				break;
+			}
+		}
+		return frames;
+	}
+
+	public Framedata translateSingleFrame( ByteBuffer buffer ) throws IncompleteException , InvalidDataException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 351-380 into a separate method.